### ISSUE TITLE
Issues with Unturned Dedicated Server

### ISSUE THREAD

**Issue by Stokaboter** [SmartlyDressedGames/Unturned-3.x-Community#4789]
Hello,

This issue is regarding +LanServer, This also might be user error so be aware of that.
I've set up my Unturned server to Buak, aside from that, no workshop.
The server launch paramaters were set in the server folder in the batch file, (Steam/steamapps/common/U3DS/ServerHelper.bat)
This was done on 2 computers without internet connection
Upon launching ServerHelper, there was no problem up until all assets have been loaded, then it tries reaching steam servers, it prompts the message "Failed to connect to Steam servers because k_EResultNoConnection, still retrying"
It doesn't make sense for the LAN server to be attempting connection to Steam servers, especially as there is no connection.
I've then set up internet for both PC's temporarily, at which point connection to the server was possible through the Lan IP (192.168.x.x)
At this point, both PCs are on the server, able to send messages through chat, interact with the enviroment, kill zombies and whatnot, 
Then, I disabled internet, the server continued to function, all functions were still there, however after about 30 minutes, the non-host PC user would get kicked for Inactivity, and cannot reconnect until internet is re-enabled.

After reconnecting, I kept Internet enabled, strangely enough, the same issue persists, even though we're both on a wired Ethernet connection, both have access to Steam Servers (Internet) it still disconnects after about 30 minutes.

I don't think there are any other parameters that require mention, however if there are, please leave a comment and I'll try to recreate/test.

**Comment by SDGNelson** [comment]
Sorry for the trouble with offline-only play! Over the past couple of years I've gotten a better appreciation for the importance of supporting offline-only functionality, and I expect it will be a core tenet of all my future projects.

There is an `-OfflineOnly` command-line flag for the server, but it has some caveats. It turns off that Steam connection requirement, and can't download workshop files. The default Steam Networking also requires an Internet connection, so a different network transport is needed, and that's where things break down somewhat... You can add `-NetTransport=SystemSockets` to use a partially-supported TCP transport, but it hasn't been worked in recently, and IIRC there's an issue when disconnecting. I'm not sure I've heard of anyone actually using it, but perhaps I can get some time soon to improve it a bit.

In short, the game is very tightly coupled with the Steam API. ;)

Note: Singleplayer should work 100% in Steam offline mode.

**Comment by Stokaboter** [comment]
Thank you for the response, and yes, the game does work perfectly in Steam offline mode, with and without workshop.
I'm not sure if i understood this correctly, but, using -OfflineOnly prevents all connections through steam, which includes Lan connections, so without some tweaking/rewriting, lan servers without internet are a nogo?.

**Comment by SDGNelson** [comment]
It should be *possible* to play on LAN without Internet access if the server has both the `-OfflineOnly` and `-NetTransport=SystemSockets` options enabled.

When I mentioned the Steam connection, I meant that by default the game server attempts to "log in" to Steam, and that's why your server was unable to start - the login was failing. (I believe it's analogous to the "login anonymous" command in SteamCMD, and if you have a GSLT it's logging into a "game server account.") With `-OfflineOnly` it *shouldn't* wait for login.

**Comment by Stokaboter** [comment]
Currently busy, so I will try -OfflineOnly and -NetTransport=SystemSockets later, but that still doesn't explain the disconnect problem we had with and without internet, unless you've addressed it somewhere and I didn't notice.

**Comment by Stokaboter** [comment]
On the server console it said that the played was kicked for inactivity.

**Comment by SDGNelson** [comment]
That would suggest the server didn't receive a packet from the player for ~30+ seconds. There might be some additional details in the server log file: https://support.smartlydressedgames.com/hc/en-us/articles/12298951221908-Where-are-the-dedicated-server-log-files

**Comment by Stokaboter** [comment]
[Server_Default_-batchmode_-nographics_Prev.log](https://github.com/user-attachments/files/17837797/Server_Default_-batchmode_-nographics_Prev.log)
Here's the full log, the first disconnect was without internet, and every disconnect after was with internet, I see all disconnects happened because a message wasn't received for ~30+ seconds, but we were next to each other and actively looting and killing zombies, so it doesn't really make sense.

**Comment by DanielWillett** [comment]
Small sidenote, this has kinda been a problem for a while but is there anything you can do about the `+WhateverServer/Name` having to be at the end of the arguments (otherwise it will include them like his did)? Maybe if the server folder doesn't exist already for the old way, use just the part up to the next argument?

I see people do that all the time (including server hosting companies).

I can make an issue if you want.

**Comment by SDGNelson** [comment]
@Stokaboter Weird - I guess let's just see what happens when/if you get the chance to try with the options mentioned.

@DanielWillett Yeah I was surprised to see that! I would typically be concerned about backwards compatibility, but your suggestion to check for an existing save with that ID is a good workaround. That said, I wonder if this is a big enough issue to warrant risking breaking something?
